pkg/app/app.go

```diff
@@ -1418,20 +1418,31 @@ Do you really want to apply?
 	// Traverse DAG of all the releases so that we don't suffer from false-positive missing dependencies
 	st.Releases = selectedAndNeededReleases
 
-	if !interactive || interactive && r.askForConfirmation(confMsg) {
```

Comment:
@tarrow Hey! Thanks for your contribution. I was trying to understand the original issue and your change fully but had no luck so far unfortunately.. .so it would be great if you could clarify a bit!

In short, can't we just inject something like the below here?

```
if len(releasesToBeUpdated) == 0 && len(releasesToBeDeleted) == 0 {
   return true, false, nil
}
```

(basically L1498-1501)

That way, I think we can fix the original issue without changing the behavior of preapply hooks.

I might have missed something but my mind says preapply hooks should run immediately before applying. Running it before prompting breaks that rule so I'm slightly inclined to not change it's behaviour this way.

BTW, do you have any motivation that you want some hooks to run even before prompting? We could add a new hook point for that, perhaps something like `preprompt`. Alterntively, you could just use `prepare` if you want some hooks to be run regardless of

Reply:
Hi @mumoshu! Thanks for the reply; sorry for the slow response.

> In short, can't we just inject something like the below here?

Yes! I think we can; I'd be very happy with this solution if you are. I have uploaded an updated patch with this change.

For clarity: I don't use any of the pre-apply hooks. I was actually trying to minimise behaviour change.

I noticed in https://github.com/helmfile/helmfile/pull/522 you said that
> `preapply` is supposed to get triggered even on unmodified releases

I was trying to keep this behaviour. With this new patchset I believe that this will no longer be true. `preapply` will now not be triggered if there are no unmodified releases. I added this to the commit message any updated the docs. Please do double check that you're happy with this since I don't understand if this change is desired (since I don't use preapply hooks).